### PR TITLE
fix NoProof block signature

### PR DESCRIPTION
**Pull request by winsvega**
Other devs report that It is better if NoProof blocks still have the signature. but the signature is just ignored. 
Rather then absence of signature as it is now.

### CODE REVIEW

File: test/tools/libtesteth/BlockChainHelper.cpp [hunk 1]
```diff
@@ -273,6 +273,14 @@ void TestBlock::mine(TestBlockChain const& _bc)
     }
     else
         recalcBlockHeaderBytes();
+
+    // Create dummy signature for NoProof blocks
+    if (Ethash::mixHash(m_blockHeader) == h256(0))
```
Comment (chfast) [inline test/tools/libtesteth/BlockChainHelper.cpp]: `h256{0}`.

File: test/tools/libtesteth/BlockChainHelper.cpp [hunk 2]
```diff
@@ -273,6 +273,14 @@ void TestBlock::mine(TestBlockChain const& _bc)
...
+    {
+        m_blockHeader.setSeal(Ethash::NonceField, h64(42));
```
Comment (chfast) [inline test/tools/libtesteth/BlockChainHelper.cpp]: Maybe make it null?

File: test/tools/libtesteth/BlockChainHelper.cpp [hunk 3]
```diff
@@ -273,6 +273,14 @@ void TestBlock::mine(TestBlockChain const& _bc)
     }
     else
         recalcBlockHeaderBytes();
+
+    // Create dummy signature for NoProof blocks
+    if (Ethash::mixHash(m_blockHeader) == h256{0})
+    {
+        m_blockHeader.setSeal(Ethash::NonceField, h64{0});
+        m_blockHeader.setSeal(Ethash::MixHashField, h256{0});
```
Comment (gumb0) [inline test/tools/libtesteth/BlockChainHelper.cpp]: Why do you assign it here if you just checked that it's 0?

Comment (winsvega) [reply test/tools/libtesteth/BlockChainHelper.cpp]: because it returns 0 also when there is no signature set.

Comment (winsvega) [reply test/tools/libtesteth/BlockChainHelper.cpp]: We could implenent a method hasSignature() if you woukd like. That will change libethereum. By default I try not to touch libethereum for test related logic

File: test/tools/jsontests/BlockChainTests.cpp [hunk 4]
```diff
@@ -109,7 +109,7 @@ json_spirit::mValue BlockchainTestSuite::doTests(json_spirit::mValue const& _inp
                     set<string> netlist;
                     json_spirit::mObject const& expectObj = expect.get_obj();
                     ImportTest::parseJsonStrValueIntoSet(expectObj.at("network"), netlist);
-
+                    netlist = test::translateNetworks(netlist);
```
Comment (gumb0) [inline test/tools/jsontests/BlockChainTests.cpp]: Did you just sneak in this unrelated change? :)

**Comment by gumb0** [comment]
I don't get - if there's a change to `NoProof` blocks format - why only testeth files are changed?
Will `test_mineBlocks` generate correct blocks with `NoProof` config?

**Comment by gumb0** [comment]
Do you mean both blocks witthout any signature and blocks with zero signature are "correct"?
This doesn't make sense to me, only one format should be valid (the one that other clients would also support)

**Comment by winsvega** [comment]
Correct in what context? 
blocks without signature are incorrect for the test format. 
blocks with or without signature could still be invalid or correct depending on a block contents. 
blocks without signature or with invalid signature are incorret on Ethash

**Comment by winsvega** [comment]
There is no consensus on blocks without signature cause NoProof is smth that we use for testing purposes. So we define what a block on NoProof should look like (with zero signature / random signature  or no signature at all) 
It is also possible to support blocks without signature in test format. It just that other clients devs agreed that it is easier for them to parse blocks with at least some signature.

**Comment by gumb0** [comment]
> So we define what a block on NoProof should look like (with zero signature / random signature or no signature at all)

So I understood that intention of this PR is to establish that NoProof blocks have zero signature.

> It just that other clients devs agreed that it is easier for them to parse blocks with at least some signature.

Right, so I'm saying we should generate blocks with signature everywhere.

> blocks with or without signature could still be invalid or correct depending on a block contents.

Why should blocks without signature be valid?

**Comment by winsvega** [comment]
Blocks without signature are valid on NoProof

**Comment by gumb0** [comment]
But you say they are not valid for other clients

**Comment by winsvega** [comment]
Other clients parsing state test format expect signature in blocks. Only in test format. They dont have NoProof defenition. We could make it anything.

**Comment by gumb0** [comment]
We could, but why would we want two different block formats in tests and out of the tests? This will only lead to confusion and possible incompatibilities between parts of the codebase.

**Comment by chfast** [comment]
> Thats why this pr adding a zero signature to the blocks without signature

But you don't change NoProof design, just hack the final test output.

I'm also a bit into the solution to make NoProof more strict and do not allow block headers without signature fields.

**Comment by winsvega** [comment]
By default I try not to touch core libraries logic for testing purposes. 
So If you think that NoProof should be modified on a core level to require signature at the block but not verifying it we could open another PR. 

To summarize the discussion: 
Is it concluded that NoProof should generate 0 signature at corelib mining method? 
Is it concluded that aleth should require signature to be present in a block rlp for any Ethash/NoProof

**Comment by chfast** [comment]
Yes, if that makes the whole system more consistent and simple to understand.

Can someone check if Clique is adding/remove any fields to the block hash?

**Comment by gumb0** [comment]
@chfast It does not change the format, just changes the meaning of some fields https://github.com/ethereum/EIPs/issues/225

**Comment by winsvega** [comment]
updatet this PR. now 0 signature is set by default inside sealEngineBase generateSeal method. 
that is used when NoProof is enabled.

**Comment by gumb0** [comment]
`bcBasicInsert` test is failing

File: libethcore/SealEngine.h [hunk 5]
```diff
@@ -99,8 +99,11 @@ class SealEngineBase: public SealEngineFace
 public:
 	void generateSeal(BlockHeader const& _bi) override
 	{
+        BlockHeader header(_bi);
```
Comment (gumb0) [inline libethcore/SealEngine.h]: I think this should happen in `NoProof`, not in `SealEngineBase`
Also I don't see any reason for `SealEngineBase` to have `generateSeal` implementation at all, so try to just move it to `NoProof`

**Comment by winsvega** [comment]
removed generateSeal from SealEngineBase
moved generateSeal from SealEngineBase to NoProof.
Fixed bcBasicImport by moving the old implementation of generateSeal from SealEngineBase to BasicAuthority generateSeal

There should be an error throwing in case no callback is called from BasicAuthority::generateSeal. 
That leads to endless loop.

**Comment by gumb0** [comment]
Where is the endless loop?

File: libethcore/SealEngine.cpp [hunk 6]
```diff
@@ -29,6 +29,17 @@ void NoProof::init()
     ETH_REGISTER_SEAL_ENGINE(NoProof);
 }
 
+void NoProof::generateSeal(BlockHeader const& _bi)
+{
+    BlockHeader header(_bi);
+    header.setSeal(1, h64{0});   // NonceField
```
Comment (gumb0) [inline libethcore/SealEngine.cpp]: Maybe move `enum { MixHashField = 0, NonceField = 1 };` to `SealEngineBase` and use it here

**Comment by winsvega** [comment]
Endless loop somewhere where it waits for a callback of m_onSealGenerated

File: libethcore/SealEngine.cpp (hunk 6 above)
Comment (winsvega) [reply libethcore/SealEngine.cpp]: this is a good question. what if with sealEngine = Ethash2  there will be 3 or 1 fields for signature?

Comment (gumb0) [reply libethcore/SealEngine.cpp]: We'll see then, let's not try to guess the future

File: libethcore/BasicAuthority.cpp [hunk 7]
```diff
@@ -50,8 +50,12 @@ void BasicAuthority::generateSeal(BlockHeader const& _bi)
...
 	h256 h = bi.hash(WithoutSeal);
 	Signature s = sign(m_secret, h);
-	setSig(bi, s);
-	SealEngineBase::generateSeal(bi);
+    setSig(bi, s);
+    RLPStream ret;
+    bi.streamRLP(ret);
+    if (m_onSealGenerated)
+        m_onSealGenerated(ret.out());
+    // Else throw an error!!!
```
Comment (gumb0) [inline libethcore/BasicAuthority.cpp]: I'd say it's a problem of the calling code if it would not set this handler and then wait somehow (how?) for it to be called.
So I don't think anything need to be done here.
